Add hook annotation to output hook logs to client on error

Bez625 opened the pull request:
Signed-off-by: Chris Berry <CBE44@ukc02zxr8zmd6m.bskyb.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Make sure to read the Contributing Guide before submitting your PR: https://github.com/helm/helm/blob/main/CONTRIBUTING.md
2. If this PR closes another issue, add 'closes #<issue number>' somewhere in the PR summary. GitHub will automatically close that issue when this PR gets merged. Alternatively, adding 'refs #<issue number>' will not close the issue, but help provide the reviewer more context.-->

**What this PR does / why we need it**:
- Implements https://github.com/helm/community/pull/301

Addresses #3481 and #2298 in outputting the logs of hook jobs and pods on failure.

Happy to add test coverage and documentation, just wanted early feedback of whether this approach would be acceptable.

**Special notes for your reviewer**:
* Using the existing c.Log(...) to leverage the --debug flag can't really work here. We could read the container logs through a buffer, but c.Log adds a newline, which would split the log lines.
* There doesn't seem to be a clear place to add tests for this, but we'll find somewhere if the general approach looks ok.

**If applicable**:
- [x] this PR contains documentation - https://github.com/helm/helm-www/pull/1242
- [x] this PR contains unit tests
- [x] this PR has been tested for backwards compatibility

arnonki commented:
It would be nice if we will have additional hook for getting the hook job/pod logs without a condition (not only on error).
Would you consider?

Bez625 commented:
> It would be nice if we will have additional hook for getting the hook job/pod logs without a condition (not only on error). Would you consider?

Hey @arnonki, we might be able to change the `ShouldOutputLogs` field to work similar to `DeletePolicies` e.g. take a comma separated list of `hook-succeeded,hook-failed`, but to be honest we have been waiting on feedback from the maintainers on whether it is worth us proceeding with this approach at all. 

We don't really want to invest time and effort writing tests and documentation if the idea is going to be rejected. 

We have tried reaching out to them on the Kubernetes slack and have now sent a mail to the cncf.io mailing list. If you like you could contribute to that mailing list discussion on helm cncf list https://lists.cncf.io/g/cncf-helm.

bacongobbler commented on pkg/kube/client.go:
```diff
@@ -620,6 +621,41 @@ func (c *Client) waitForPodSuccess(obj runtime.Object, name string) (bool, error
 
+// GetPodList uses the build in
+func (c *Client) GetPodList(namespace string, listOptions metav1.ListOptions) (*v1.PodList, error) {
+	podList, err := c.kubeClient.CoreV1().Pods(namespace).List(context.Background(), listOptions)
```
What if users explicitly set the `metadata.namespace` parameter in their pods to some other namespace? We've seen that happen quite often with a few charts. Restricting to the namespace means we cannot fetch pods managed by the chart deployed to another namespace IIRC.

z4ce replied:
It looks like this has been addressed now.. the deriveNamespace function is determining what gets passed in here.

bacongobbler commented on pkg/kube/client.go:
```diff
@@ -620,6 +621,41 @@ func (c *Client) waitForPodSuccess(obj runtime.Object, name string) (bool, error
 	return false, nil
 }
 
+// GetPodList uses the build in
```
incomplete docs

bacongobbler commented:
Hi @Bez625. A couple of quick notes.

I noticed no test coverage was provided with these changes. That makes it much more difficult for the maintainers to review as we don't know what's the expected behaviour of the code. Is it a bug or a feature? We strive for 80% test coverage.

I don't see any documentation surrounding how to test this feature ("if I do X, I should expect to see Y"), how users are expected to use this (how do I set this up?), and how it is intended to be used (and in some cases, what scenarios it is not intended for use).

Those two things alone make it incredibly difficult to gauge how to evaluate this feature. As it stands now all I see is some code without much context.

Addressing those concerns may help present your use case to the maintainers for review. By making it easier for us to review, we can provide quicker feedback on your proposal.

Hope that helps.

bacongobbler commented on pkg/kube/fake/printer.go:
```diff
@@ -101,6 +103,16 @@ func (p *PrintingKubeClient) WaitAndGetCompletedPodPhase(_ string, _ time.Durati
 	return v1.PodSucceeded, nil
 }
 
+// GetPodList implements KubeClient GetPodListForSelector.
```
typo?

bacongobbler commented on pkg/releaseutil/manifest_sorter.go:
```diff
@@ -231,3 +233,13 @@ func operateAnnotationValues(entry SimpleHead, annotation string, operate func(p
+// shouldLogOutput
+func shouldLogOutput(entry SimpleHead) bool {
+	if outputLogs, ok := entry.Metadata.Annotations[release.HookOutputLogAnnotation]; ok {
+		if outputLogs == "true" {
```
What about "yes", or other truthy values? What about "True" or "TRUE" or "tRuE"? Shouldn't those evaluate as true?

Bez625 commented:
> Hi @Bez625. A couple of quick notes.
> 
> I noticed no test coverage was provided with these changes. That makes it much more difficult for the maintainers to review as we don't know what's the expected behaviour of the code. Is it a bug or a feature? We strive for 80% test coverage.
> 
> I don't see any documentation surrounding how to test this feature ("if I do X, I should expect to see Y"), how users are expected to use this (how do I set this up?), and how it is intended to be used (and in some cases, what scenarios it is not intended for use).
> 
> Those two things alone make it incredibly difficult to evaluate this feature. As it stands, all I see is some code without much context.
> 
> Addressing those concerns may help present your use case to the maintainers for review. By making it easier for us to review, we can provide quicker feedback on your proposal.
> 
> Hope that helps.

Hi Matt,

Those are fair comments - I think we were only hoping for a quick sense check at this early stage to see if it was worth continuing. It sounds like you are receptive to the idea and there is obviously some other interest in the feature so we will proceed with docs and testings. 

We will also address your other comments.

Thanks a lot for the feedback
Chris

haarchri commented:
we looking for this feature to get log output from "flux check --pre"  https://github.com/fluxcd-community/helm-charts/issues/18

Bez625 commented:
Hi @bacongobbler! We have added the testing and associated docs to the PR. Please could you review and let us know if anything needs to be changed? 🙏

Bez625 commented:
Hi @bacongobbler is there anything else we need to do to progress this? Thanks!

bacongobbler commented:
Hi @Bez625. Just getting back from the holidays. Apologies for the delay on a response.

I am stepping away from day-to-day code reviews for the Helm project, choosing to focus more on [reviewing new HIPs](https://github.com/helm/community/blob/main/hips/hip-0001.md), [onboarding](https://github.com/helm/helm/pull/10326) new maintainers [to the project](https://github.com/helm/community/blob/main/hips/hip-0014.md), and [reviewing incoming security audits](https://github.com/helm/community/blob/main/SECURITY.md). As a result, I will not have enough time to come back to this.

Given this PR is labeled a `size/L`, it will require approval from two core maintainers [as per the contributing policy](https://github.com/helm/helm/blob/main/CONTRIBUTING.md#size-labels). I assume everyone else was out for the holidays so I'd try asking again in a week for a review once everyone is back and all settled in.

Thanks for being patient.

Bez625 commented:
> Hi @Bez625. Just getting back from the holidays. Apologies for the delay on a response.
> 
> I am stepping away from day-to-day code reviews for the Helm project, choosing to focus more on [reviewing new HIPs](https://github.com/helm/community/blob/main/hips/hip-0001.md), [onboarding](https://github.com/helm/helm/pull/10326) new maintainers [to the project](https://github.com/helm/community/blob/main/hips/hip-0014.md), and [reviewing incoming security audits](https://github.com/helm/community/blob/main/SECURITY.md). As a result, I will not have enough time to come back to this.
> 
> Given this PR is labeled a `size/L`, it will require approval from two core maintainers [as per the contributing policy](https://github.com/helm/helm/blob/main/CONTRIBUTING.md#size-labels). I assume everyone else was out for the holidays so I'd try asking again in a week for a review once everyone is back and all settled in.
> 
> Thanks for being patient.

Thanks for replying @bacongobbler - is there anyone in particular we should be reaching out to for reviews or a github group we can tag if there is no traction within a week?

bridgetkromhout commented:
>  is there anyone in particular we should be reaching out to for reviews or a github group we can tag if there is no traction within a week?

Hi @Bez625, reaching out to specific maintainers probably isn't the approach that will be the most helpful. I added this to the agenda for the [upcoming weekly community call](https://github.com/helm/community/blob/main/communication.md#meetings), which you are welcome to attend.

However, most maintainers who have cycles available this week are likely working on the upcoming 3.8 release right now: https://github.com/helm/helm/milestone/113. 

If we can get your PR reviewed in the coming weeks, then it may be possible to add it to an upcoming milestone. Thanks!

technosophos commented on pkg/kube/interface.go:
```diff
@@ -68,6 +70,12 @@ type Interface interface {
 	IsReachable() error
+
+	// GetPodList list all pods that match the specified listOptions
+	GetPodList(namespace string, listOptions metav1.ListOptions) (*v1.PodList, error)
```
Adding new methods to an interface is considered a breaking change. So this PR would either need to refactor this or defer it to Helm 4.

technosophos replied:
Just to be clearer -- I was looking at the above while getting ready for the discussion on the Helm call. Sorry for the late comment. But, yes, adding methods to a public interface is considered a breaking change, and thus has to be deferred to a major release. However, I think this one should be easy enough to refactor into a separate interface and some assertions about interface type. (The Go codebase itself does this quite a bit).

Bez625 replied:
Hi @technosophos - I have refactored the new methods on to the `InterfaceExt` interface similarly to `WaitForDelete`. I'd hope this is sufficient and we can move all breaking change methods to `Interface` as part of Helm 4. I have added a TODO next to the type assertions as they should be able to be removed also once methods are moved to Interface. Please let me know if any issues, thanks again!

bridgetkromhout replied:
Thanks for the refactor - we should be able to consider this now.

Bez625 replied:
Thanks very much @bridgetkromhout !

mattfarina replied:
I'm sorry to say that the Go guidelines say that extending ANY interface is a breaking change. You need to create a new Interface for this. Something like `InterfaceLogs`. The Go article that we link in the Helm backwards compatibility rules is at https://go.dev/blog/module-compatibility#working-with-interfaces. This describes what to do.

Bez625 replied:
Hi Matt, apologies i've just seen this. I thought the entire point of `InterfaceExt` was to act as a temporary extension until the changes could be merged in to `Interface` in helm4 from the doc string:

```golang
// InterfaceExt is introduced to avoid breaking backwards compatibility for Interface implementers.
//
// TODO Helm 4: Remove InterfaceExt and integrate its method(s) into the Interface.
```

scottrigby replied:
continuing thread here: https://github.com/helm/helm/pull/10309#pullrequestreview-2547572308

Bez625 committed:
Add hook annotations to output pod logs to client on success and fail

Signed-off-by: Chris Berry <bez625@gmail.com>

mkorejo commented:
will this enhancement allow users to view hook logs regardless of error or not?

we have a helm pre-install/pre-upgrade hook which runs DB schema migrations, and I'm wondering if this enhancement will allow me to easily view the pod logs?

Bez625 commented:
> will this enhancement allow users to view hook logs regardless of error or not?
> 
> we have a helm pre-install/pre-upgrade hook which runs DB schema migrations, and I'm wondering if this enhancement will allow me to easily view the pod logs?

Yes the output log policy is designed to be configurable. There are more details in the accompanying docs PR [here](https://github.com/helm/helm-www/pull/1242/files#diff-0415a9abbbcb28820b6339e175864f4ec055f9222050b0f5af492d611d0a2b75R217)

eabykov commented:
@Bez625 @bacongobbler what about if the `pod` fell by mistake during helm upgrade/install?

Examples of the `helm_upgrade_logs` scripts:
- https://github.com/SamuelGarrattIqa/helm_upgrade_logs/blob/main/bin/helm_upgrade_logs.sh
- https://github.com/felipecrs/dotfiles/blob/master/home/dot_local/bin/executable_helm-upgrade-logs.tmpl

Example of GitHub actions: https://github.com/SamuelGarrattIqa/helm_upgrade_logs/actions/runs/2923030002/jobs/4661178723

Bez625 commented:
> @Bez625 @bacongobbler what about if the `pod` fell by mistake during helm upgrade/install?
> 
> Examples of the `helm_upgrade_logs` scripts:
> 
> * https://github.com/SamuelGarrattIqa/helm_upgrade_logs/blob/main/bin/helm_upgrade_logs.sh
> * https://github.com/felipecrs/dotfiles/blob/master/home/dot_local/bin/executable_helm-upgrade-logs.tmpl
> 
> Example of GitHub actions: https://github.com/SamuelGarrattIqa/helm_upgrade_logs/actions/runs/2923030002/jobs/4661178723

I'm not entirely sure I understand the question. The intention of the feature is to be able to get the output from pods or jobs created as pre or post hooks and send it to the helm client. This is configurable for either success or failure or both. 

What kind of mistake do you mean?

eabykov commented:
> > @Bez625 @bacongobbler what about if the `pod` fell by mistake during helm upgrade/install?
> > Examples of the `helm_upgrade_logs` scripts:
> > 
> > * https://github.com/SamuelGarrattIqa/helm_upgrade_logs/blob/main/bin/helm_upgrade_logs.sh
> > * https://github.com/felipecrs/dotfiles/blob/master/home/dot_local/bin/executable_helm-upgrade-logs.tmpl
> > 
> > Example of GitHub actions: https://github.com/SamuelGarrattIqa/helm_upgrade_logs/actions/runs/2923030002/jobs/4661178723
> 
> I'm not entirely sure I understand the question. The intention of the feature is to be able to get the output from pods or jobs created as pre or post hooks and send it to the helm client. This is configurable for either success or failure or both.
> 
> What kind of mistake do you mean?

I mean the case when we don't have hooks and we just run command `helm install/upgrade`, a pod is created and we wait for it to be ready, but if it crashes with any error N times and we don't have time for the timeout, then we roll back the release. I would like to see pod shell crash logs in the console and k8s events in real time if it possible

Bez625 commented:
> I mean the case when we don't have hooks and we just run command `helm install/upgrade`, a pod is created and we wait for it to be ready, but if it crashes with any error N times and we don't have time for the timeout, then we roll back the release. I would like to see pod shell crash logs in the console and k8s events in real time if it possible

This PR isnt intended for that particular case and is specific to hooks. However, you could potentially leverage this functionality by having a pre install hook that runs as a job checking the status/logs of your target deployment.

Bez625 commented:
@bridgetkromhout @technosophos is there any update to this? It's been waiting over a year now!

Ornias1993 commented:
> @bridgetkromhout @technosophos is there any update to this? It's been waiting over a year now!

In general the maintainers here basically ignore feature addition PR's.

salzig commented:
> > @bridgetkromhout @technosophos is there any update to this? It's been waiting over a year now!
> 
> In general the maintainers here basically ignore feature addition PR's.

that's very very sad. 

We would love to see this feature, so we could implement a pre-rollback that prints out some helpful hints for debugging.

z4ce commented:
Hey @Bez625  I brought up this PR on the developer call today. Are you still interested in this PR?

I'm looking for some guidance from @mattfarina but if things look good, would you be willing to rebase?

z4ce commented on pkg/kube/client.go:
```diff
@@ -68,7 +70,7 @@ type Client struct {
 	Namespace string
 
-	kubeClient *kubernetes.Clientset
+	kubeClient kubernetes.Interface
```
I think this would have be reworked to not change the Client struct as it maintainers guide says `Go interfaces in pkg/ MUST NOT be modified`. Is that right?

z4ce commented:
I have a branch here where I've merged this patch with the current main branch of helm https://github.com/z4ce/helm/commits/output_merged

Bez625 commented:
> Hey @Bez625 I brought up this PR on the developer call today. Are you still interested in this PR?
> 
> I'm looking for some guidance from @mattfarina but if things look good, would you be willing to rebase?

Apologies I completely missed this. I'm happy to rebase if it is going to be reviewed, would you like to me to do so?

Bez625 commented:
> I have a branch here where I've merged this patch with the current main branch of helm https://github.com/z4ce/helm/commits/output_merged

I have updated the fork and conflicts look resolved. Please let me know if there is anything else you need.

Bez625 committed:
Tidy up imports

Signed-off-by: Chris <bez625@gmail.com>

joejulian reviewed (dismissed):
This looks good to me and works for some test cases I have. Approving pending HIP merge.

Bez625 commented:
> This looks good to me and works for some test cases I have. Approving pending HIP merge.

Thanks @joejulian! Is there anything else we need to do to get this merged?

sergelogvinov commented:
@mattfarina @z4ce 
Hello, I saw your discussion on YouTube, and I'd like to emphasize the importance of PR.

This feature is typically uses in CI/CD pipelines like Jenkins, TeamCity and GitHub Actions. When deployments fail, developers or QA engineers often resort to checking the CI/CD logs on the CI interface. It's the most effective way to show them the reason for the deployment failure.

Flux already supports notifications, and I believe they can also integrate helm-output into the notification body.

Thank you.

gusdan commented:
Thank for that work. We really need it!!!

LuBingtan commented:
any updates?

mattfarina commented:
@Bez625 Can you please rebase and fix the conflicts.

Bez625 committed:
Fix lint

Signed-off-by: Chris Berry <bez625@gmail.com>

Bez625 commented:
> @Bez625 Can you please rebase and fix the conflicts.

Hi @mattfarina - I have rebased and fixed conflicts. I also replied to comment https://github.com/helm/helm/pull/10309/files#r1735244931

Bez625 commented:
Hey @mattfarina - anything else you need me to do?

Would be nice to try and get this merged, possibly even before the 3 year anniversary

scottrigby reviewed (changes requested):
@Bez625 thanks for all the work that went into this PR. The HIP, the functionality implementation here, and especially the tests. Also appreciate all the feedback you've incorporated into this PR, as well as the documentation PR. There is only one issue that I believe needs to be addressed.

@technosophos and @mattfarina both noted a required change that has not yet been implemented in this PR. I understand your comment here https://github.com/helm/helm/pull/10309/files#r1735244931, and I believe I understand where the confusion came from. I believe the comment you referred to is a confusing one. 

The comment was added as part of https://github.com/helm/helm/pull/9702. I believe its intent was to explain why there was an additional interface created (to keep backwards-compatibility), and to remind us that it should be cleaned up in the next MAJOR version of Helm. However, I believe the comment is worded strangely—as if any future additions in Helm 3 should be made to this interface. That is not the case, it only explains why the previous addition was made. The fact that this comment exists does not prevent changing this interface from being a breaking change.

In short, I'd like to request you make the following changes:

1. Create a new Interface for your addition in this PR, and update the usages accordingly. As Matt suggested, something like `InterfaceLogs` or another name you feel makes sense. Please also include a similar comment as `InterfaceExt` has (with one change, see below).

2. Please also clarify the wording in the comment above `InterfaceExt` to not suggest it should be used for ongoing changes. I think changing to past tense should be sufficient:

```diff
- // InterfaceExt is introduced to avoid breaking backwards compatibility for Interface implementers.
+ // InterfaceExt was introduced to avoid breaking backwards compatibility for Interface implementers.
```

3. mark `InterfaceExt` as deprecated, and recommend using your new interface instead.

Bez625 committed:
Update based on review comments

Signed-off-by: Chris Berry <bez625@gmail.com>

Bez625 commented:
> @Bez625 thanks for all the work that went into this PR. The HIP, the functionality implementation here, and especially the tests. Also appreciate all the feedback you've incorporated into this PR, as well as the documentation PR. There is only one issue that I believe needs to be addressed.
> 
> @technosophos and @mattfarina both noted a required change that has not yet been implemented in this PR. I understand your comment here https://github.com/helm/helm/pull/10309/files#r1735244931, and I believe I understand where the confusion came from. I believe the comment you referred to is a confusing one.
> 
> The comment was added as part of #9702. I believe its intent was to explain why there was an additional interface created (to keep backwards-compatibility), and to remind us that it should be cleaned up in the next MAJOR version of Helm. However, I believe the comment is worded strangely—as if any future additions in Helm 3 should be made to this interface. That is not the case, it only explains why the previous addition was made. The fact that this comment exists does not prevent changing this interface from being a breaking change.
> 
> In short, I'd like to request you make the following changes:
> 
> 1. Create a new Interface for your addition in this PR, and update the usages accordingly. As Matt suggested, something like `InterfaceLogs` or another name you feel makes sense. Please also include a similar comment as `InterfaceExt` has (with one change, see below).
> 2. Please also clarify the wording in the comment above `InterfaceExt` to not suggest it should be used for ongoing changes. I think changing to past tense should be sufficient:
> 
> ```diff
> - // InterfaceExt is introduced to avoid breaking backwards compatibility for Interface implementers.
> + // InterfaceExt was introduced to avoid breaking backwards compatibility for Interface implementers.
> ```
> 
> 3. mark `InterfaceExt` as deprecated, and recommend using your new interface instead.

Hi @scottrigby 

Thank you for your detailed reply. I have removed the methods from the existing `InterfaceExt` and created a new `InterfaceLogs`. I have left the `InterfaceExt` definition as it was and NOT deprecated, as that seems to be the pattern with the other `Interface*` interfaces that have been introduced. I've also added `was` to my interface doc string as it seems the policy is that once an "extension" interface is merged it cannot be changed. 

Please let me know if any more changes are required.

Thanks

Bez625 committed:
Update based on review comments

Signed-off-by: Chris Berry <bez625@gmail.com>

scottrigby reviewed (dismissed):
Nice work 👍 

> I have left the InterfaceExt definition as it was and NOT deprecated

This is fine. My thought was that you could replace the other instances if you wanted, but that's not necessary—how you've done works just as well.

> I've also added was to my interface doc string as it seems the policy is that once an "extension" interface is merged it cannot be changed.

Just to be clear here, changing after _merge_ is not necessarily a problem, it's only a problem to change after a _release_ within the same MAJOR version, in order to fulfill Helm's strict backwards compatibility in accordance with Go's non-breaking changes. 

I should note that very recently `main` is actually for Helm 4 now, but you were so close to v3 backwards compatibility I didn't want to have you reverse all that. This last change has been helpful so that a backport to `dev-v3` will be less painful (as the window for that for new features is drawing to a close). Helm 4 will be getting lots of cleanup in the near future and simplifying these interfaces can be a near-future follow up as part of that cleanup. So again, thank you.

> Please let me know if any more changes are required.

Requested changes were made, tests still look accurate and pass. I think you're good now 👍

Bez625 commented:
> Nice work 👍
> 
> > I have left the InterfaceExt definition as it was and NOT deprecated
> 
> This is fine. My thought was that you could replace the other instances if you wanted, but that's not necessary—how you've done works just as well.
> 
> > I've also added was to my interface doc string as it seems the policy is that once an "extension" interface is merged it cannot be changed.
> 
> Just to be clear here, changing after _merge_ is not necessarily a problem, it's only a problem to change after a _release_ within the same MAJOR version, in order to fulfill Helm's strict backwards compatibility in accordance with Go's non-breaking changes.
> 
> I should note that very recently `main` is actually for Helm 4 now, but you were so close to v3 backwards compatibility I didn't want to have you reverse all that. This last change has been helpful so that a backport to `dev-v3` will be less painful (as the window for that for new features is drawing to a close). Helm 4 will be getting lots of cleanup in the near future and simplifying these interfaces can be a near-future follow up as part of that cleanup. So again, thank you.
> 
> > Please let me know if any more changes are required.
> 
> Requested changes were made, tests still look accurate and pass. I think you're good now 👍

Awesome, thanks @scottrigby . Do we need to get @mattfarina to approve or can we just dismiss his request for changes since they have been made now?

scottrigby commented:
> Awesome, thanks @scottrigby . Do we need to get @mattfarina to approve or can we just dismiss his request for changes since they have been made now?

We need one more maintainer's approval. I raised at today's Helm dev meeting. We just have a pretty big backlog for reviews, but some folks said they'd take a look.

Matt's concern has been addressed, so we could dismiss the change request. But will keep for now in case Matt is the 2nd maintainer review in that case he can do that himself.

scottrigby commented:
Update: https://github.com/helm/community/pull/301 has been merged as [hip-0019](https://github.com/helm/community/blob/main/hips/hip-0019.md).

Bez625 commented:
Thanks @scottrigby - please let me know if there is anything else I need to do and I'll keep an eye for second reviewer feedback.

scottrigby commented:
@Bez625 PR needs rebase due to go mod/sum now conflicting. This will happen periodically until this PR is merged.

Bez625 commented:
@scottrigby I have rebased

scottrigby reviewed (approved):
registering my approval again, after rebase/stale review was dismissed. See https://github.com/helm/helm/pull/10309#pullrequestreview-2551465286

gjenkins8 commented on pkg/action/hooks.go:
```diff
@@ -87,20 +94,32 @@ func (cfg *Configuration) execHook(rl *release.Release, hook release.HookEvent,
 			// under failed condition. If so, then clear the corresponding resource object in the hook
-			if err := cfg.deleteHookByPolicy(h, release.HookFailed, timeout); err != nil {
-				return err
+			if errDeleting := cfg.deleteHookByPolicy(h, release.HookFailed, timeout); err != nil {
```
Is this meant to be?
```suggestion
			if errDeleting := cfg.deleteHookByPolicy(h, release.HookFailed, timeout); errDeleting != nil {
```

scottrigby replied:
You're right, this is a bug and your suggestion fixes it.

gjenkins8 commented on pkg/action/hooks.go:
```diff
@@ -158,3 +177,62 @@ func hookHasDeletePolicy(h *release.Hook, policy release.HookDeletePolicy) bool
+}
+
+func (cfg *Configuration) deriveNamespace(h *release.Hook, namespace string) (string, error) {
+	values, err := chartutil.ReadValues([]byte(h.Manifest))
```
I don't think chart's Values type should be used to process generic yaml...

A generic parser should be used instead. e.g. https://pkg.go.dev/k8s.io/apimachinery/pkg/runtime/serializer#CodecFactory.UniversalDeserializer

Bez625 replied:
I have used `yaml.Unmarshal` with a temp struct to extract the namespace. I took a look at the UniversalDeserializer, but I couldnt find a generic type to satisfy both jobs/pods and it felt like it required a lot of configuration for a simple usecase. 

Please let me know if there is a better approach.

gjenkins8 commented on pkg/action/hooks.go:
```diff
@@ -158,3 +177,62 @@ func hookHasDeletePolicy(h *release.Hook, policy release.HookDeletePolicy) bool
+// hookHasOutputLogPolicy determines whether the defined hook output log policy matches the hook output log policies
+// supported by helm.
+func hookHasOutputLogPolicy(h *release.Hook, policy release.HookOutputLogPolicy) bool {
+	for _, v := range h.OutputLogPolicies {
```
nit: https://pkg.go.dev/slices#Contains could be used now

gjenkins8 commented on pkg/action/hooks.go:
```diff
@@ -158,3 +177,62 @@ func hookHasDeletePolicy(h *release.Hook, policy release.HookDeletePolicy) bool
+		}
+		switch h.Kind {
+		case "Job":
+			return cfg.outputContainerLogsForListOptions(namespace, metav1.ListOptions{LabelSelector: fmt.Sprintf("job-name=%s", h.Name)})
```
Unless I am mistaken, this is requiring jobs to have a `job-name=<hook name>` label for output to work

scottrigby replied:
I had to dig a little, but this is correct as-is. The release.Hook.Name is the Helm release name.

gjenkins8 commented on pkg/kube/interface.go:
```diff
@@ -72,19 +72,30 @@ type Interface interface {
+	GetPodList(namespace string, listOptions metav1.ListOptions) (*v1.PodList, error)
+
+	// OutputContainerLogsForPodList output the logs for a pod list
+	OutputContainerLogsForPodList(podList *v1.PodList, namespace string, writer io.Writer) error
```
https://github.com/helm/community/blob/main/hips/hip-0019.md calls for a `HookOutputFunc`. Here we are writing to the logger directly

Bez625 replied:
I missed this in the HIP that @z4ce created. I have tried to integrate the `HookOutputFunc` in the way I thought it was intended.

gjenkins8 reviewed (changes requested):
Some questions, and probably a bug or two.  And need to align a few things with  https://github.com/helm/community/blob/main/hips/hip-0019.md I think

scottrigby commented on pkg/action/hooks.go:
```diff
@@ -87,20 +94,32 @@ func (cfg *Configuration) execHook(rl *release.Release, hook release.HookEvent,
-				return err
+			if errDeleting := cfg.deleteHookByPolicy(h, release.HookFailed, timeout); err != nil {
+				// We log the error here as we want to propagate the hook failure upwards to the release object.
+				// This is a change in behaviour as the edge case previously would lose the hook error and only
```
fair

scottrigby replied:
fixed in 16079d5

scottrigby reviewed (changes requested):
I was also experimenting more with this locally—and found this same issue before I saw @gjenkins8's comment https://github.com/helm/helm/pull/10309#discussion_r1938370584 At the least, this needs to be fixed before this PR is ready

scottrigby commented on pkg/action/hooks.go:
```diff
@@ -87,20 +94,32 @@ func (cfg *Configuration) execHook(rl *release.Release, hook release.HookEvent,
+				// We log the error here as we want to propagate the hook failure upwards to the release object.
+				// This is a change in behaviour as the edge case previously would lose the hook error and only
+				// raise the delete hook error.
+				log.Printf("error the hook resource on hook failure: %v", errDeleting)
```
I think it would make more sense to add "deleting" to this error message:
```suggestion
				log.Printf("error deleting the hook resource on hook failure: %v", errDeleting)
```

scottrigby commented on pkg/action/hooks.go:
```diff
@@ -87,20 +94,32 @@ func (cfg *Configuration) execHook(rl *release.Release, hook release.HookEvent,
-				return err
+			if errDeleting := cfg.deleteHookByPolicy(h, release.HookFailed, timeout); err != nil {
+				// We log the error here as we want to propagate the hook failure upwards to the release object.
+				// This is a change in behaviour as the edge case previously would lose the hook error and only
+				// raise the delete hook error.
```
To fix https://github.com/helm/helm/pull/10309#discussion_r1938366884 just remove these two comment lines:

```suggestion
```

scottrigby committed:
add short circuit return

Co-authored-by: George Jenkins <gvjenkins@gmail.com>
Signed-off-by: Scott Rigby <scott@r6by.com>

scottrigby committed:
remove comments about previous functionality

Signed-off-by: Scott Rigby <scott@r6by.com>

scottrigby committed:
fix err check

Co-authored-by: George Jenkins <gvjenkins@gmail.com>
Signed-off-by: Scott Rigby <scott@r6by.com>

scottrigby committed:
clarify fix error message

Signed-off-by: Scott Rigby <scott@r6by.com>

Bez625 committed:
Add HookOutputFunc and generic yaml unmarshaller

Signed-off-by: Chris Berry <bez625@gmail.com>

Bez625 commented:
@scottrigby @gjenkins8 - I have addressed the review comments and rebased. Apologies I was unable to look at this until today.

banjoh commented on pkg/action/action.go:
```diff
@@ -422,6 +427,12 @@ func (cfg *Configuration) Init(getter genericclioptions.RESTClientGetter, namesp
 	cfg.KubeClient = kc
 	cfg.Releases = store
 	cfg.Log = log
+	cfg.HookOutputFunc = defaultHookOutputWriter
```
> Additionally, there will be a new item added to the action SDK configuration to allow SDK consumers to get the output. By default this output will be discarded by writing to io.Discard, but an SDK consumer can overwrite the HookOutputFunc to provide a custom writer.

Snippet from the HIP.

We need to allow SDK users to provide their own `HookOutputFunc`. The default would be one that writes to `io.Discard`

Bez625 replied:
I made changes to the Init method signature before I saw your second comment. 

How do SDK users normally interact with the `Configuration` struct? I assumed they would build their own rather than use the `Init`...

Assuming we can't change the `Init` signature and we can't default to a function returning `iog.Writer()` then the only option is to set the `cfg.HookOutputFunc` after the `Init` func has run. I can add a utility to do this for now, but if you have a preferred alternative please let me know.

banjoh replied:
https://helm.sh/docs/sdk/examples/ will have examples of how an SDK user would use Helm

banjoh replied:
We should be able to change the public interfaces, be we cannot backport this PR to v3 helm.

Bez625 replied:
Ok - given this PR no longer changes the existing interfaces I think we should be good to have this in v3. 

Looking at the SDK examples I think the current solution of setting the `HookOutputFunc` in the `main` function after the `Init` has run should work.

banjoh replied:
> I can add a utility to do this for now, but if you have a preferred alternative please let me know.

This could work as wel.

scottrigby replied:
@Bez625 Nice job 👍 This works! Thanks for @banjoh for clarifying @gjenkins8 's comment https://github.com/helm/helm/pull/10309#discussion_r1938374610 that it needed to be user-configurable. It is now, so I think we are good on this.

Bez625 committed:
Refactor based on review comment

Signed-off-by: Chris Berry <bez625@gmail.com>

Bez625 committed:
Refactor based on review comment

Signed-off-by: Chris Berry <bez625@gmail.com>

scottrigby reviewed (approved):
@Bez625 thanks for your persistence on this 3+ year old PR. You've addressed all maintainer concerns, provided tests which pass, and manual testing I've done again still works. Approving!

> Additionally, a new user flag should be created --no-log-output that would skip the output of logs.

HIP NOTE: The flag in this note on the HIP was not added in this PR. Given where we are with the functionality, I think this is good for the main point of the HIP. I believe the flag is a lower priority in the HIP, and can be added in a follow-up PR.

BACKPORT NOTE: You have also successfully added this feature in a way that does not add any breaking changes if we were to be able to backport this to Helm 3 (this will be based on available time as we're focusing on Helm 4, so no promises there, but thank you for doing that too, and we'll keep you posted).

gjenkins8 reviewed (approved):
Thanks for the PR! And for the changes (after a lengthy wait). Appreciate the contribution
